// bs-fetch/src/Fetch.res.mjs


import * as Curry from "rescript/lib/es6/curry.js";
import * as Caml_option from "rescript/lib/es6/caml_option.js";

function encodeRequestMethod(method_) {
  if (typeof method_ === "object") {
    return method_._0;
  }
  switch (method_) {
    case "Get" :
        return "GET";
    case "Head" :
        return "HEAD";
    case "Post" :
        return "POST";
    case "Put" :
        return "PUT";
    case "Delete" :
        return "DELETE";
    case "Connect" :
        return "CONNECT";
    case "Options" :
        return "OPTIONS";
    case "Trace" :
        return "TRACE";
    case "Patch" :
        return "PATCH";
    
  }
}

function encodeReferrerPolicy(param) {
  switch (param) {
    case "None" :
        return "";
    case "NoReferrer" :
        return "no-referrer";
    case "NoReferrerWhenDowngrade" :
        return "no-referrer-when-downgrade";
    case "SameOrigin" :
        return "same-origin";
    case "Origin" :
        return "origin";
    case "StrictOrigin" :
        return "strict-origin";
    case "OriginWhenCrossOrigin" :
        return "origin-when-cross-origin";
    case "StrictOriginWhenCrossOrigin" :
        return "strict-origin-when-cross-origin";
    case "UnsafeUrl" :
        return "unsafe-url";
    
  }
}

function encodeRequestMode(param) {
  switch (param) {
    case "Navigate" :
        return "navigate";
    case "SameOrigin" :
        return "same-origin";
    case "NoCORS" :
        return "no-cors";
    case "CORS" :
        return "cors";
    
  }
}

function encodeRequestCredentials(param) {
  switch (param) {
    case "Omit" :
        return "omit";
    case "SameOrigin" :
        return "same-origin";
    case "Include" :
        return "include";
    
  }
}

function encodeRequestCache(param) {
  switch (param) {
    case "Default" :
        return "default";
    case "NoStore" :
        return "no-store";
    case "Reload" :
        return "reload";
    case "NoCache" :
        return "no-cache";
    case "ForceCache" :
        return "force-cache";
    case "OnlyIfCached" :
        return "only-if-cached";
    
  }
}

function encodeRequestRedirect(param) {
  switch (param) {
    case "Follow" :
        return "follow";
    case "Error" :
        return "error";
    case "Manual" :
        return "manual";
    
  }
}

var HeadersInit = {};

function append(arg1, arg2, obj) {
  obj.append(arg1, arg2);
}

function $$delete(arg1, obj) {
  obj.delete(arg1);
}

function get(arg1, obj) {
  return Caml_option.null_to_opt(obj.get(arg1));
}

function has(arg1, obj) {
  return obj.has(arg1);
}

function set(arg1, arg2, obj) {
  obj.set(arg1, arg2);
}

var $$Headers = {
  append: append,
  $$delete: $$delete,
  get: get,
  has: has,
  set: set
};

var BodyInit = {};

function map(f, v) {
  if (v !== undefined) {
    return Caml_option.some(Curry._1(f, Caml_option.valFromOption(v)));
  }
  
}

function make(method_, headers, body, referrer, referrerPolicyOpt, mode, credentials, cache, redirect, integrityOpt, keepalive, signal) {
  var referrerPolicy = referrerPolicyOpt !== undefined ? referrerPolicyOpt : "None";
  var integrity = integrityOpt !== undefined ? integrityOpt : "";
  var partial_arg = integrity;
  var partial_arg$1 = map(encodeRequestRedirect, redirect);
  var partial_arg$2 = map(encodeRequestCache, cache);
  var partial_arg$3 = map(encodeRequestCredentials, credentials);
  var partial_arg$4 = map(encodeRequestMode, mode);
  var partial_arg$5 = encodeReferrerPolicy(referrerPolicy);
  var partial_arg$6 = map(encodeRequestMethod, method_);
  return function (param) {
    var tmp = {};
    if (partial_arg$6 !== undefined) {
      tmp._method = partial_arg$6;
    }
    if (headers !== undefined) {
      tmp.headers = Caml_option.valFromOption(headers);
    }
    if (body !== undefined) {
      tmp.body = Caml_option.valFromOption(body);
    }
    if (referrer !== undefined) {
      tmp.referrer = referrer;
    }
    if (partial_arg$5 !== undefined) {
      tmp.referrerPolicy = partial_arg$5;
    }
    if (partial_arg$4 !== undefined) {
      tmp.mode = partial_arg$4;
    }
    if (partial_arg$3 !== undefined) {
      tmp.credentials = partial_arg$3;
    }
    if (partial_arg$2 !== undefined) {
      tmp.cache = partial_arg$2;
    }
    if (partial_arg$1 !== undefined) {
      tmp.redirect = partial_arg$1;
    }
    if (partial_arg !== undefined) {
      tmp.integrity = partial_arg;
    }
    if (keepalive !== undefined) {
      tmp.keepalive = keepalive;
    }
    if (signal !== undefined) {
      tmp.signal = Caml_option.valFromOption(signal);
    }
    return tmp;
  };
}

function method_(self) {
  var method_$1 = self.method;
  switch (method_$1) {
    case "CONNECT" :
        return "Connect";
    case "DELETE" :
        return "Delete";
    case "GET" :
        return "Get";
    case "HEAD" :
        return "Head";
    case "OPTIONS" :
        return "Options";
    case "PATCH" :
        return "Patch";
    case "POST" :
        return "Post";
    case "PUT" :
        return "Put";
    case "TRACE" :
        return "Trace";
    default:
      return {
              TAG: "Other",
              _0: method_$1
            };
  }
}

function type_(self) {
  var e = self.type;
  switch (e) {
    case "" :
        return "None";
    case "audio" :
        return "Audio";
    case "font" :
        return "Font";
    case "image" :
        return "Image";
    case "script" :
        return "Script";
    case "style" :
        return "Style";
    case "track" :
        return "Track";
    case "video" :
        return "Video";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestType: " + e,
            Error: new Error()
          };
  }
}

function destination(self) {
  var e = self.destination;
  switch (e) {
    case "" :
        return "None";
    case "document" :
        return "Document";
    case "embed" :
        return "Embed";
    case "font" :
        return "Font";
    case "image" :
        return "Image";
    case "manifest" :
        return "Manifest";
    case "media" :
        return "Media";
    case "object" :
        return "Object";
    case "report" :
        return "Report";
    case "script" :
        return "Script";
    case "serviceworker" :
        return "ServiceWorker";
    case "sharedworder" :
        return "SharedWorker";
    case "style" :
        return "Style";
    case "worker" :
        return "Worker";
    case "xslt" :
        return "Xslt";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestDestination: " + e,
            Error: new Error()
          };
  }
}

function referrerPolicy(self) {
  var e = self.referrerPolicy;
  switch (e) {
    case "" :
        return "None";
    case "no-referrer" :
        return "NoReferrer";
    case "no-referrer-when-downgrade" :
        return "NoReferrerWhenDowngrade";
    case "origin" :
        return "Origin";
    case "origin-when-cross-origin" :
        return "OriginWhenCrossOrigin";
    case "same-origin" :
        return "SameOrigin";
    case "strict-origin" :
        return "StrictOrigin";
    case "strict-origin-when-cross-origin" :
        return "StrictOriginWhenCrossOrigin";
    case "unsafe-url" :
        return "UnsafeUrl";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown referrerPolicy: " + e,
            Error: new Error()
          };
  }
}

function mode(self) {
  var e = self.mode;
  switch (e) {
    case "cors" :
        return "CORS";
    case "navigate" :
        return "Navigate";
    case "no-cors" :
        return "NoCORS";
    case "same-origin" :
        return "SameOrigin";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestMode: " + e,
            Error: new Error()
          };
  }
}

function credentials(self) {
  var e = self.credentials;
  switch (e) {
    case "include" :
        return "Include";
    case "omit" :
        return "Omit";
    case "same-origin" :
        return "SameOrigin";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestCredentials: " + e,
            Error: new Error()
          };
  }
}

function cache(self) {
  var e = self.cache;
  switch (e) {
    case "default" :
        return "Default";
    case "force-cache" :
        return "ForceCache";
    case "no-cache" :
        return "NoCache";
    case "no-store" :
        return "NoStore";
    case "only-if-cached" :
        return "OnlyIfCached";
    case "reload" :
        return "Reload";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestCache: " + e,
            Error: new Error()
          };
  }
}

function redirect(self) {
  var e = self.redirect;
  switch (e) {
    case "error" :
        return "Error";
    case "follow" :
        return "Follow";
    case "manual" :
        return "Manual";
    default:
      throw {
            RE_EXN_ID: "Failure",
            _1: "Unknown requestRedirect: " + e,
            Error: new Error()
          };
  }
}

function classify(t) {
  if (typeof t === "string") {
    return {
            NAME: "String",
            VAL: t
          };
  } else {
    return {
            NAME: "File",
            VAL: t
          };
  }
}

var EntryValue = {
  classify: classify
};

function append$1(arg1, arg2, obj) {
  obj.append(arg1, arg2);
}

function $$delete$1(arg1, obj) {
  obj.delete(arg1);
}

function get$1(arg1, obj) {
  return obj.get(arg1);
}

function getAll(arg1, obj) {
  return obj.getAll(arg1);
}

function set$1(arg1, arg2, obj) {
  obj.set(arg1, arg2);
}

function has$1(arg1, obj) {
  return obj.has(arg1);
}

function appendObject(arg1, arg2, filename, obj) {
  obj.append(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

function appendBlob(arg1, arg2, filename, obj) {
  obj.append(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

function appendFile(arg1, arg2, filename, obj) {
  obj.append(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

function setObject(arg1, arg2, filename, obj) {
  obj.set(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

function setBlob(arg1, arg2, filename, obj) {
  obj.set(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

function setFile(arg1, arg2, filename, obj) {
  obj.set(arg1, arg2, filename !== undefined ? Caml_option.valFromOption(filename) : undefined);
}

var $$FormData = {
  EntryValue: EntryValue,
  Iterator: undefined,
  append: append$1,
  $$delete: $$delete$1,
  get: get$1,
  getAll: getAll,
  set: set$1,
  has: has$1,
  appendObject: appendObject,
  appendBlob: appendBlob,
  appendFile: appendFile,
  setObject: setObject,
  setBlob: setBlob,
  setFile: setFile
};

var $$AbortController = {
  abort: (function (prim) {
      prim.abort();
    })
};

function Body_arrayBuffer(prim) {
  return prim.arrayBuffer();
}

function Body_blob(prim) {
  return prim.blob();
}

function Body_formData(prim) {
  return prim.formData();
}

function Body_json(prim) {
  return prim.json();
}

function Body_text(prim) {
  return prim.text();
}

var Body = {
  arrayBuffer: Body_arrayBuffer,
  blob: Body_blob,
  formData: Body_formData,
  json: Body_json,
  text: Body_text
};

var RequestInit = {
  make: make
};

function Request_arrayBuffer(prim) {
  return prim.arrayBuffer();
}

function Request_blob(prim) {
  return prim.blob();
}

function Request_formData(prim) {
  return prim.formData();
}

function Request_json(prim) {
  return prim.json();
}

function Request_text(prim) {
  return prim.text();
}

var $$Request = {
  method_: method_,
  type_: type_,
  destination: destination,
  referrerPolicy: referrerPolicy,
  mode: mode,
  credentials: credentials,
  cache: cache,
  redirect: redirect,
  arrayBuffer: Request_arrayBuffer,
  blob: Request_blob,
  formData: Request_formData,
  json: Request_json,
  text: Request_text
};

function Response_clone(prim) {
  return prim.clone();
}

function Response_arrayBuffer(prim) {
  return prim.arrayBuffer();
}

function Response_blob(prim) {
  return prim.blob();
}

function Response_formData(prim) {
  return prim.formData();
}

function Response_json(prim) {
  return prim.json();
}

function Response_text(prim) {
  return prim.text();
}

var $$Response = {
  clone: Response_clone,
  arrayBuffer: Response_arrayBuffer,
  blob: Response_blob,
  formData: Response_formData,
  json: Response_json,
  text: Response_text
};

export {
  $$AbortController ,
  HeadersInit ,
  $$Headers ,
  BodyInit ,
  Body ,
  RequestInit ,
  $$Request ,
  $$Response ,
  $$FormData ,
}
/* No side effect */
